### PR TITLE
Implement GUB branching

### LINKED ISSUE #158: Multi variable branching
## Describe the problem or improvement suggested
Create a new environment for MVB.

## Describe the solution you would like
PySCIPOpt solution from @CharJon
```python
def branch_on_disjunction(model: scip.Model, coefficients, variables):
    """Branch on the sum of integer/binary variables.
        Always creates both children and will error out
        if branching would not cut off current solution!
        As nodeselprio and childestimate for 'createChild'
        are unknown, the node-selection-heuristic will be
        quite random.
    """
    assert len(variables) > 0, f"Need at least one variable to branch on"
    assert len(variables) == len(coefficients), f"Need as many coefficients as variables"
    assert all(v.vtype() == 'BINARY' for v in variables), f"Binary only implemented"
    current_values_sum = sum(v.getLPSol() for v in variables)
    assert not model.isFeasIntegral(current_values_sum), f"Sum {current_values_sum} is not fractional"
    floored_bound = math.floor(current_values_sum)
    if len(variables) == 1:
        model.branchVar(variables[0])
    elif floored_bound == 0:
        # trick to allow better bound checking
        child_1 = model.createChild(1, 0)
        for cur_variable in variables:
            model.chgVarUbNode(child_1, cur_variable, 0)
        child_2 = model.createChild(1, 0)
        model.addConsNode(child_2, scip.quicksum(c * v for c, v in zip(coefficients, variables)) >= 1)
    elif floored_bound + 1 == len(variables):
        # trick to allow better bound checking
        child_1 = model.createChild(1, 0)
        model.addConsNode(child_1, scip.quicksum(c * v for c, v in zip(coefficients, variables)) <= floored_bound)
        child_2 = model.createChild(1, 0)
        for cur_variable in variables:
            model.chgVarLbNode(child_2, cur_variable, 1)
    else:
        child_1 = model.createChild(1, 0)
        model.addConsNode(child_1, scip.quicksum(c * v for c, v in zip(coefficients, variables)) <= floored_bound)
        child_2 = model.createChild(1, 0)
        model.addConsNode(child_2, scip.quicksum(c * v for c, v in zip(coefficients, variables)) >= floored_bound + 1)
    return scip.SCIP_RESULT.BRANCHED
```

### PR DESCRIPTION
**Pull request by AntoinePrv**
## Pull request checklist

- [x] I have open an issue to discuss the proposed changes. Fix #158
- [x] I have modified/added tests to cover the new changes/features.
- [x] I have modified/added the documentation to cover the new changes/features.
- [x] I have ran the tests, checks, and code formatters.

## General idea
For integer variables `x_i` provided by the user, and their LP relaxation value `x_lp_i`,
 - If `s := sum(x_lp_i)` is fractional, then branch on `sum(x_i) <= floor(s)` & `sum(x_i) >= ceil(s)`
 - If `s := sum(x_lp_i)` is integer, then branch on `sum(x_i) <= s - 1` & `sum(x_i) == s` & `sum(x_i) >= s+ 1`

## Proposed implementation
 - When `n_vars == 0`, fail;
 - Otherwise (including when `n_vars == 1`):
   - Fail if any variable is continuous or fixed
   - if `sum(x_lp_i)` is infinite, what then?
   - if `sum(x_lp_i)` is integer, branch on three nodes
   - else branch on two nodes

### CODE REVIEW

**Comment by AntoinePrv** [comment]
When given all the action set
```python
_, action_set, _, done, _ = env.reset(next(instances))
while not done:
    _, action_set, _, done, _ = env.step(action_set)
```
And that their sum is integral, three nodes are created. But since no variable get fixed, they all appear again in the `action_set` and the code loops indefinitely.

File: libecole/include/ecole/dynamics/branching-gub.hpp [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+class BranchingGUBDynamics :
+	public EnvironmentDynamics<nonstd::span<std::size_t const>, std::optional<xt::xtensor<std::size_t, 1>>> {
+public:
+	using Action = nonstd::span<std::size_t const>;
+	using ActionSet = std::optional<xt::xtensor<std::size_t, 1>>;
+
+	std::tuple<bool, ActionSet> reset_dynamics(scip::Model& model) override;
+
+	std::tuple<bool, ActionSet> step_dynamics(scip::Model& model, Action const& var_idx) override;
```
Comment (gasse) [inline libecole/include/ecole/dynamics/branching-gub.hpp:21]: Maybe plural, var_idxs ?

File: libecole/src/dynamics/branching-gub.cpp [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+#include <scip/struct_var.h>
+
+namespace {
+
+SCIP_RETCODE SCIPbranchGUB_add_child(
+	SCIP* scip,
+	SCIP_Real priority,
+	SCIP_Real estimate,
+	SCIP_VAR** vars,
+	SCIP_Real* ones,
```
Comment (gasse) [inline libecole/src/dynamics/branching-gub.cpp:24]: Why is this argument called ones ? Does it have to be a vector of ones ? If so, why is it an argument ?

File: python/src/ecole/core/dynamics.cpp [hunk 3]
```diff
@@ -15,25 +15,253 @@ namespace ecole::dynamics {
...
+					The state of the Markov Decision Process. Passed by the environment.
+				action:
+					The index the LP column of the variable to branch on. One element of the action set.
+
+			Returns
+			-------
+				done:
+					Whether the instance is solved.
+				action_set:
+					List of branching canddidates. Candidates depend on parameters in :py:meth:`__init__`.
```
Comment (gasse) [inline python/src/ecole/core/dynamics.cpp:107]: typo canddidates -> candidates

File: python/src/ecole/core/dynamics.cpp [hunk 4]
```diff
@@ -15,25 +15,253 @@ namespace ecole::dynamics {
...
+				)")
+		.def_set_dynamics_random_state(R"(
+			Set seeds on the :py:class:`~ecole.scip.Model`.
+
+			Set seed parameters, inculing permutation, LP, and shift.
```
Comment (gasse) [inline python/src/ecole/core/dynamics.cpp:112]: typo inculing -> including

File: python/src/ecole/core/dynamics.cpp [hunk 5]
```diff
@@ -15,25 +15,253 @@ namespace ecole::dynamics {
...
+		)");
+
+	using idx_t = typename BranchingGUBDynamics::Action::value_type;
+	using array_t = py::array_t<idx_t, py::array::c_style | py::array::forcecast>;
+	dynamics_class<BranchingGUBDynamics>{m, "BranchingGUBDynamics", R"(
+		Global Upper Bound branching Dynamics.
+
+		Based on a SCIP branching callback with maximal priority and no depth limit.
+		The dynamics give the control back to the user every time the callback would be called.
+		The user recieves as an action set the list of branching candiates, and is expected to select
```
Comment (gasse) [inline python/src/ecole/core/dynamics.cpp:138]: typo candiates -> candidates

File: python/src/ecole/core/dynamics.cpp [hunk 6]
```diff
@@ -15,25 +15,253 @@ namespace ecole::dynamics {
...
+				done:
+					Whether the instance is solved.
+					This can happen without branching, for instance if the instance is solved during presolving.
+				action_set:
+					List of branching candidates (``SCIPgetPseudoBranchCands``).
+		)")
+		.def_set_dynamics_random_state(R"(
+			Set seeds on the :py:class:`~ecole.scip.Model`.
+
+			Set seed parameters, inculing permutation, LP, and shift.
```
Comment (gasse) [inline python/src/ecole/core/dynamics.cpp:169]: typo inculing

File: python/src/ecole/core/dynamics.cpp [hunk 7]
```diff
@@ -15,25 +15,253 @@ namespace ecole::dynamics {
...
+				auto const vars = nonstd::span{action.data(), static_cast<std::size_t>(action.size())};
+				auto const release = py::gil_scoped_release{};
+				return self.step_dynamics(model, vars);
+			},
+			py::arg("model"),
+			py::arg("action"),
+			R"(
+			Branch and resume solving until next branching.
+
+			Branching is done on a the sum of the given variables using their LP or pseudo solution value.
```
Comment (gasse) [inline python/src/ecole/core/dynamics.cpp:190]: typo a the sum

File: python/src/ecole/core/dynamics.cpp [hunk 8]
```diff
@@ -15,25 +15,253 @@ namespace ecole::dynamics {
...
+			must be non integer.
+			In the opposite case, an exception will be thrown,
+			The control is given back to the user on the next branching decision or when done.
+
+			Parameters
+			----------
+				model:
+					The state of the Markov Decision Process. Passed by the environment.
+				action:
+					A subset of of the variables of given in the aaction set.
```
Comment (gasse) [inline python/src/ecole/core/dynamics.cpp:201]: typo aaction

File: python/src/ecole/core/dynamics.cpp [hunk 9]
```diff
@@ -15,25 +15,253 @@ namespace ecole::dynamics {
...
+					Not all subsets are valid.
```
Comment (gasse) [inline python/src/ecole/core/dynamics.cpp:202]: maybe add (see above)

File: python/src/ecole/core/dynamics.cpp [hunk 10]
```diff
@@ -15,25 +15,253 @@ namespace ecole::dynamics {
...
+					Whether the instance is solved.
+				action_set:
+					List of branching candidates (``SCIPgetPseudoBranchCands``).
+		)")
+		.def(py::init<>());
+
+	dynamics_class<ConfiguringDynamics>{m, "ConfiguringDynamics", R"(
+		Setting solving parameters dynamics.
+
+		This dynamics are meant to be used as a (contextual) bandit to find good parameters for SCIP.
```
Comment (gasse) [inline python/src/ecole/core/dynamics.cpp:216]: is meant ?

File: python/src/ecole/core/dynamics.cpp [hunk 11]
```diff
@@ -15,25 +15,253 @@ namespace ecole::dynamics {
...
+			-------
+				done:
+					Whether the instance is solved. Always true.
+				action_set:
+					Unused.
+		)")
+		.def_set_dynamics_random_state(R"(
+			Set seeds on the :py:class:`~ecole.scip.Model`.
+
+			Set seed parameters, inculing permutation, LP, and shift.
```
Comment (gasse) [inline python/src/ecole/core/dynamics.cpp:256]: typo inculing

File: python/src/ecole/core/dynamics.cpp [hunk 12]
```diff
@@ -15,25 +15,253 @@ namespace ecole::dynamics {
...
 
-	dynamics_class<ConfiguringDynamics>(m, "ConfiguringDynamics")  //
+			Parameters
+			----------
+				model:
+					The state of the Markov Decision Process. Passed by the environment.
+				random_engine:
+					The source of randomness.Passed by the environment.
```
Comment (gasse) [inline python/src/ecole/core/dynamics.cpp:263]: missing space randomness.Passed

**Review by gasse — APPROVED** [review]
Minor comments about variable names and a few typos. High quality PR !

**Pull request merged**